### PR TITLE
feat(ingest): add Feast repository source

### PR DESCRIPTION
**Pull request by danilopeixoto**
This pull request adds Feast repository source.

- Support latest Feast version
- Remove Docker and simplify ingestion (the challenge is here)
- Preserve existing Feast source

I'm opening this as a suggestion and I would appreciate feedback on other implementation strategies.

**How does it work?**

1. The team clones the Feast repository (eg. local, JupyterLab, or CI pipeline environment).
2. The team injects variables into the environment to use those in the Feast project.
3. Apply feature metadata:

```console
feast apply
```

```yaml
# feature_store.yaml

registry:
  path: 's3://${FEAST_S3_BUCKET_NAME}/metadata.db'
project: 'demo'
provider: 'local'
online_store:
  type: 'redis'
  redis_type: 'redis'
  connection_string: '${REDIS_HOSTNAME}:${REDIS_PORT},password=${REDIS_PASSWORD}'
offline_store:
  type: 'file'
flags:
  alpha_features: true
  direct_ingest_to_online_store: true
  on_demand_transforms: true
```

4. Sync feature metadata:

```console
datahub ingest -c feast_to_datahub_recipe.yaml
```

```yaml
# feast_to_datahub_recipe.yaml

source:
  type: 'feast-repository'
  config:
    path: '.'
    environment: 'PROD'
sink:
  type: 'datahub-rest'
  config:
    server: ${DATAHUB_SERVER}
    token: ${DATAHUB_TOKEN}
```

## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/source_docs/feast_repository.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  # sink configs
+```
+
+## Config details
+
+Note that a `.` is used to denote nested fields in the YAML recipe.
+
+| Field         | Required | Default | Description                                |
+| ------------- | -------- | ------- | ------------------------------------------ |
+| `path`        | `yes`    |         | Path to Feast repository.                  |
```
Comment (kevinhu) [inline metadata-ingestion/source_docs/feast_repository.md:43]: nit: for consistency, could you use checkmark icons for "Required"? ex: https://datahubproject.io/docs/metadata-ingestion/source_docs/postgres/#config-details

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    ValueType.UNIX_TIMESTAMP: MLFeatureDataType.TIME,
+    ValueType.BYTES_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.STRING_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.INT32_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.INT64_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.DOUBLE_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.FLOAT_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.BOOL_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.UNIX_TIMESTAMP_LIST: MLFeatureDataType.SEQUENCE,
+    ValueType.NULL: MLFeatureDataType.USELESS,
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:55]: The USELESS type should be for unique data with no meaningful use — for instance, UUIDs. The `UNKNOWN` type would be a better fit.

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        feature_sources = []
+
+        if isinstance(feature_view, FeatureView):
+            feature_sources = self._get_data_sources(project, feature_view)
+        elif isinstance(feature_view, OnDemandFeatureView):
+            if feature_view.input_request_data_sources is not None:
+                for (
+                    _,
+                    request_source,
+                ) in feature_view.input_request_data_sources.items():
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:208]: If the key is unused here, could we use `.values()` instead?

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+                    feature_sources.append(
+                        builder.make_dataset_urn(
+                            source_platform,
+                            f"{project}.{source_name}",
+                            self.source_config.environment,
+                        )
+                    )
+
+            if feature_view.input_feature_views is not None:
+                for _, feature_view_source in feature_view.input_feature_views.items():
```
Comment (kevinhu) [inline metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:222]: If the key is unused here, could we use `.values()` instead?

**Review by kevinhu — COMMENTED** [review]
Looks great! Just a few little comments—thanks for authoring this!

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py (hunk 4 above)
Comment (danilopeixoto) [reply metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:222]: Thanks for finding it!

File: metadata-ingestion/source_docs/feast_repository.md (hunk 1 above)
Comment (danilopeixoto) [reply metadata-ingestion/source_docs/feast_repository.md]: Done! We lost alignment in raw, but that's ok!
image

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py (hunk 2 above)
Comment (danilopeixoto) [reply metadata-ingestion/src/datahub/ingestion/source/feast_repository.py]: Fixed! I was reading this: `NULL` as `USELESS` value for model inputs.

**Comment by shirshanka** [comment]
@danilopeixoto : Does it make sense to have a single feast source instead of two different sources?

**Comment by danilopeixoto** [comment]
> @danilopeixoto : Does it make sense to have a single feast source instead of two different sources?

Yes! I'm with you! But Feast has changed the design since 0.10 version. More details here: https://docs.feast.dev/project/feast-0.9-vs-feast-0.10+. Now, it is modular, CLI-centric and based on repositories: https://docs.feast.dev/how-to-guides/running-feast-in-production.

As Feast has two different designs, it was easier to release another source.

**Comment by kevinhu** [comment]
@danilopeixoto It would be useful if we could clarify the distinction between the two versions—how about adding a sentence or two in the documentation for both feast and feast-repository informing the user about when to use the other one?

Also, please add this source to https://datahubproject.io/docs/cli/#sources

File: metadata-ingestion/source_docs/feast_repository.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+```
+
+## Config details
+
+Note that a `.` is used to denote nested fields in the YAML recipe.
+
+| Field         | Required | Default | Description                                |
+| ------------- | -------- | ------- | ------------------------------------------ |
+| `path`        | ✅       |         | Path to Feast repository.                  |
+| `environment` |          | `PROD`  | Environment to use when constructing URNs. |
```
Comment (rslanka) [inline metadata-ingestion/source_docs/feast_repository.md:44]: @danilopeixoto, does this source need any credentials to connect?

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+
+class FeastRepositorySourceConfig(ConfigModel):
+    path: str
+    environment: str = DEFAULT_ENV
+
+
+class FeastRepositorySource(Source):
+    source_config: FeastRepositorySourceConfig
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:65]: Please move these from class-scope to instance-scope (to __init__).

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        super().__init__(ctx)
+
+        self.source_config = config
+
+    def _get_field_type(self, field_type: str, parent_name: str) -> str:
+        """
+        Maps types encountered in Feast to corresponding schema types.
+        """
+
+        enum_type = _field_type_mapping.get(field_type)
```
Comment (rslanka) [inline metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:78]: rename `enum_type` to `ml_feature_data_type`?

Comment (danilopeixoto) [reply metadata-ingestion/src/datahub/ingestion/source/feast_repository.py]: Done!

File: metadata-ingestion/src/datahub/ingestion/source/feast_repository.py (hunk 6 above)
Comment (danilopeixoto) [reply metadata-ingestion/src/datahub/ingestion/source/feast_repository.py:65]: Looking for other sources, they always keep the configuration model in class-scope (eg. [MySQLConfig](https://github.com/danilopeixoto/datahub/blob/97a61c36964bf3579722d06b45f5baec110437f0/metadata-ingestion/src/datahub/ingestion/source/sql/mysql.py#L28)).

Would this change break consistency?

File: metadata-ingestion/source_docs/feast_repository.md (hunk 5 above)
Comment (danilopeixoto) [reply metadata-ingestion/source_docs/feast_repository.md:44]: Yes! But at Feast level. How does it work?

1. The team clones the Feast repository (eg. local, JupyterLab, or CI pipeline environment).
2. The team injects variables into the environment to use them in the Feast project definition:

```yaml
# feature_store.yaml

registry:
    path: 's3://${FEAST_S3_BUCKET_NAME}/metadata.db'
project: 'demo'
provider: 'local'
online_store:
    type: 'redis'
    redis_type: 'redis'
    connection_string: '${REDIS_HOSTNAME}:${REDIS_PORT},password=${REDIS_PASSWORD}'
offline_store:
    type: 'file'
flags:
    alpha_features: true
    direct_ingest_to_online_store: true
    on_demand_transforms: true
```

If the Feast CLI and SDK are working for the chosen offline/online/registry stores:

```console
feast apply
```

then DataHub source will also work:

```console
datahub ingest -c feast_to_datahub_recipe.yaml
```

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 8]
```diff
@@ -25,283 +37,334 @@
...
+        """
+        Generate an MLPrimaryKey work unit for a Feast entity.
+        """
+
+        feature_view_name = f"{self.feature_store.project}.{feature_view.name}"
+
+        entity_snapshot = MLPrimaryKeySnapshot(
+            urn=builder.make_ml_primary_key_urn(feature_view_name, entity.name),
+            aspects=[StatusClass(removed=False)],
+        )
```
Comment (achals) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:179]: I'm not super familiar with Datahub conceptually, but from the Feast side an entity can be the "primary key" into multiple different feature views. Is `MLPrimaryKeySnapshot` specific to a `MLFeatureTable`, in this case a FeatureView?

Comment (gabe-lyons) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:179]: Hey @achals - MlPrimaryKey can be associated with multiple MlFeatureTables. The association happens here:

https://github.com/datahub-project/datahub/blob/master/metadata-models/src/main/pegasus/com/linkedin/ml/metadata/MLFeatureTableProperties.pdl#L51

**Review by shirshanka — APPROVED** [review]
LGTM! CI failures look unrelated. 
Thanks for the patience on this @danilopeixoto!

**Pull request merged**